src/java.base/share/classes/sun/security/util/Cache.java

```diff
@@ -248,33 +267,45 @@ public void accept(CacheVisitor<K,V> visitor) {
 
 class MemoryCache<K,V> extends Cache<K,V> {
 
-    private static final float LOAD_FACTOR = 0.75f;
-
-    // XXXX
+    // Debugging
     private static final boolean DEBUG = false;
 
     private final Map<K, CacheEntry<K,V>> cacheMap;
     private int maxSize;
+    final private int maxQueueSize;
     private long lifetime;
     private long nextExpirationTime = Long.MAX_VALUE;
+    // Locking is to protect QueueCacheEntry's from being removed from the
+    // cacheMap while another thread is adding new queue entries.
+    private ReentrantLock lock;
```

Comment:
Every method that accesses the cacheMap needs to be synchronized using the same synchronization method. As it stands now, `clear` and `remove` are synchronized, `get` uses the lock, `put` uses both, and `expungeExpiredEntries` is sometimes called without any synchronization.

Note that reads (`get`, `size`) must be synchronized with writes. There are reports or `get` ending up in an infinite loop when called without synchronization.

I recommended reverting everything to `synchronized` because that makes the PR smaller. I guess you could replace all uses of `synchronized` with a ReentrantLock, but I don't see any benefit in doing that.